File: factory/src/main/java/com/zuzex/factory/callback/KafkaCarStatusSendCallback.java
package com.zuzex.factory.callback;

import com.zuzex.common.callback.BaseKafkaSendCallback;
import com.zuzex.common.dto.CarStatusDto;
import com.zuzex.factory.model.Order;
import com.zuzex.factory.service.OrderService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.springframework.context.annotation.Lazy;
import org.springframework.kafka.core.KafkaProducerException;
import org.springframework.lang.NonNull;
import org.springframework.stereotype.Component;

@Slf4j
@Component
@RequiredArgsConstructor
public class KafkaCarStatusSendCallback extends BaseKafkaSendCallback<String, CarStatusDto> {

    @Lazy
    private final OrderService orderService;

    @Override
    public void onFailure(@NonNull KafkaProducerException ex) {
        log.info("Message '{}' failed for reason: {}", ex.getFailedProducerRecord().value(), ex.getMessage());
        ProducerRecord<String, CarStatusDto> failedProducerRecord = ex.getFailedProducerRecord();
        Long invalidOrderId = failedProducerRecord.value().getCarId();

        Order orderAfterSave = orderService.revertOrderStatusById(invalidOrderId);
        log.info("Consistency returned: {}", orderAfterSave);
    }
}

File: car-factory/src/main/kotlin/com/zuzex/carfactory/callback/KafkaCarStatusSendCallback.kt
package com.zuzex.carfactory.callback

import com.zuzex.carfactory.service.OrderService
import com.zuzex.common.callback.BaseKafkaSendCallback
import com.zuzex.common.dto.CarStatusDto
import mu.KotlinLogging
import org.springframework.context.annotation.Lazy
import org.springframework.kafka.core.KafkaProducerException
import org.springframework.lang.NonNull
import org.springframework.stereotype.Component

private val log = KotlinLogging.logger {}

@Component
class KafkaCarStatusSendCallback(@Lazy private val orderService: OrderService) :
    BaseKafkaSendCallback<String, CarStatusDto>() {

    override fun onFailure(@NonNull ex: KafkaProducerException) {
        val failedProducerRecord = ex.getFailedProducerRecord<String, CarStatusDto>().value()
        log.info("Message '$failedProducerRecord' failed for reason: ${ex.message}")
        orderService.revertOrderStatusById(failedProducerRecord.carId)
            .also { log.info("Consistency returned: $it") }
    }
}
